### PR TITLE
Add PDF generation ability to Microsoft.DocAsCode.App

### PR DESCRIPTION
**Pull request by dpvreony**
Greetings

* Adds the ability to target the pdf config section in docfx.json
* Adds logging around templates not being resolved, as came across an issue which had no logging and results in 'No TOC file is included, no PDF file will be generated'

partial resolution of #8872

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/TemplateProcessors/TemplateProcessor.cs [hunk 1]
```diff
@@ -35,6 +35,10 @@ public TemplateProcessor(ResourceFileReader resourceProvider, DocumentBuildConte
         _resourceProvider = resourceProvider;
         _maxParallelism = maxParallelism;
         _templateCollection = new TemplateCollection(resourceProvider, context, maxParallelism);
+        if (_templateCollection.Count == 0)
+        {
+            Logger.LogWarning("No template bundles were found, no template will be applied to the documents. 1) Check your docfx.json 2) or your working directory has the templates subfolder");
```
Comment (filzrev) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateProcessors/TemplateProcessor.cs:40]: > 2) or your working directory has the templates subfolder

I though current docfx.exe don't use `{cwd}/templates/**` for resource lookup.

https://github.com/dotnet/docfx/blob/c1545fe6f006a41df1551ed53d6b9961bacdc4a8/src/Microsoft.DocAsCode.Build.Engine/TemplateProcessors/TemplateManager.cs#L45-L61

It seems template folder are searched with following orders.
1. Launched EXE directory (`AppContext.BaseDirectory`)
2. `_baseDirectory` that is passed to TemplateProcessor's constructor.

When RunBuild/RunPdf operation is executed. `_baseDirectory` is always set to `docfx.json` existing directory.
So. `cwd` is not used for template resource lookup.

**Comment by yufeih** [comment]
@dpvreony build failed with:

```

Error: /home/runner/work/docfx/docfx/test/docfx.Tests/DocsetTest.cs(198,30): error xUnit1013: Public method 'Build_With_RedirectUri_Files' on test class 'DocsetTest' should be marked as a Fact. Reduce the visibility of the method, or add a Fact attribute to the method. [/home/runner/work/docfx/docfx/test/docfx.Tests/docfx.Tests.csproj::TargetFramework=net6.0]
Error: /home/runner/work/docfx/docfx/test/docfx.Tests/DocsetTest.cs(198,30): error xUnit1013: Public method 'Build_With_RedirectUri_Files' on test class 'DocsetTest' should be marked as a Fact. Reduce the visibility of the method, or add a Fact attribute to the method. [/home/runner/work/docfx/docfx/test/docfx.Tests/docfx.Tests.csproj::TargetFramework=net8.0]
Error: /home/runner/work/docfx/docfx/test/docfx.Tests/DocsetTest.cs(198,30): error xUnit1013: Public method 'Build_With_RedirectUri_Files' on test class 'DocsetTest' should be marked as a Fact. Reduce the visibility of the method, or add a Fact attribute to the method. [/home/runner/work/docfx/docfx/test/docfx.Tests/docfx.Tests.csproj::TargetFramework=net7.0]
```

File: test/docfx.Tests/DocsetTest.cs [hunk 2]
```diff
@@ -217,4 +244,164 @@ public static async Task Build_With_RedirectUri_Files()
         var urls = XDocument.Parse(sitemapXml).Root.Elements();
         Assert.True(urls.Count() == 0);
     }
+
+    [Fact]
+    public static async Task Pdf_Basic()
```
Comment (yufeih) [inline test/docfx.Tests/DocsetTest.cs:249]: nit: consider move PDF related tests to a separate file.

File: test/docfx.Tests/DocsetTest.cs [hunk 3]
```diff
@@ -217,4 +244,164 @@ public static async Task Build_With_RedirectUri_Files()
...
+            ["pdf/toc.yml"] = "- name: Introduction\r\n  href: intro.md\r\n- name: Another Page\r\n  href: anotherpage.md\r\n",
+            ["pdf/cover.md"] = "# My Basic Cover Page\r\n\r\n",
+            ["pdf/intro.md"] = "# Introduction\r\n\r\n",
+            ["pdf/anotherpage.md"] = "# Another Page\r\n\r\n"
+        });
+
+        Assert.True(outputs.ContainsKey("Pdf_With_Cover_Page_pdf.pdf"));
+    }
+
+    [Fact]
```
Comment (yufeih) [inline test/docfx.Tests/DocsetTest.cs:317]: Would this PDF tests pass on linux and mac? If not, we may want to skip them using conditional facts.

**Review by yufeih — COMMENTED** [review]
Some tests that validate logs failed due to new warnings. We could either fix those tests or drop the new warning change.

File: test/docfx.Tests/DocsetTest.cs (hunk 3 above)
Comment (dpvreony) [reply test/docfx.Tests/DocsetTest.cs]: are you intending to adopt arcade? there's a reusable fact in there https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/Tests/Microsoft.NET.TestFramework/Attributes/WindowsOnlyFactAttribute.cs

Comment (yufeih) [reply test/docfx.Tests/DocsetTest.cs]: To me, it seems simpler to copy the code as there aren't a lot of lines there.

**Comment by yufeih** [comment]
@dpvreony  It seems tests are still failing.

**Review by yufeih — APPROVED** [review]

**Comment by yufeih** [comment]
Thank you @dpvreony for adding this feature and taking the time to make sure tests work!

**Pull request merged**